File: app/gui/Controller.js
var path = require('path');
var cp = require('child_process');

var MyInputStream = require('./MyInputStream');
var MyOutputStream = require('./MyOutputStream');

var Controller = function(inputHandler, errHandler) {
    // spawn the ruby process
    var cwd = process.cwd();
    var controllerPath = path.join(cwd, 'app', 'controller');
    var rubyPath = path.join(cwd, 'Ruby', 'bin', 'ruby');
    var rubyProcess = cp.spawn(rubyPath, [path.join(controllerPath, 'main.rb')], {
        cwd: controllerPath
    });
    rubyProcess.stderr.on('data', errHandler);

    this.input = new MyInputStream(rubyProcess.stdout, inputHandler);
    this.output = new MyOutputStream(rubyProcess.stdin);

    this.write = function(type, data) {
        this.output.write(type, data);
    }
    
    this.exit = function() {
        this.write('exit');
    }
    
    this.fetchInfo = function() {
        this.write('fetchInfo');
    }
    
    this.newTask = function(url, path) {
        this.write('new', {
            url: url,
            path: path
        });
    }
    
    this.manageTask = function(op, id) {
        this.write(op, {
            id: id
        });
    }
    
    this.closeSupporter = function() {
        this.write('closeSupporter');
    }
    
    this.openSupporter = function() {
        this.write('openSupporter');
    }
    
    this.connect = function(ipAddr, myIpAddr) {
        this.write('connect', {
            ipAddr: ipAddr,
            myIpAddr: myIpAddr
        });
    }
}


module.exports = Controller;

File: app/controller/Controller.rb
require './MyInputStream'
require './MyOutputStream'
require './LocalTaskManager'
require './SupporterManager'
require './MasterManager'
require './SupporterListener'
require './MasterSearcher'
require './SupporterSearcher'
require './Util'

class Controller
    include Constants
    
    def initialize
        @ltm = LocalTaskManager.new
        @stms = []
        @sm = SupporterManager.new @ltm
        @sl = SupporterListener.new @sm
        @mm = MasterManager.new
        @ms = MasterSearcher.new self
        @ss = SupporterSearcher.new
        @output = MyOutputStream.new STDOUT
        @input = MyInputStream.new STDIN, true do |type, data|
            inputHandler type, data
        end
    end

    def write type, data = {}
        @output.write type, data
    end

    def newTask path, url
        id = @ltm.newTask path, url
        @sm.newTask id, url
    end

    def startTask id
        url = @ltm.startTask id
        @sm.newTask id, url
    end

    def suspendTask id
        @sm.deleteTask id
        @ltm.suspendTask id
    end

    def deleteTask id
        @sm.deleteTask id
        @ltm.deleteTask id
    end

    def connect ipAddr
        return if @mm.include? ipAddr
        socket = TCPSocket.new ipAddr, MASTER_PORT
        @mm.newMaster socket
    end

    def sendInfo
        data = {
            'info' => {
                'tasks' => @ltm.tasksInfo,
                'supportersNum' => @sm.supportersNum,
                'supporterState' => true
            }
        }
        write 'info', data
    end

    def closeSupporter
        @ms.close
        @mm.removeAll
    end

    def finalize
        @sm.deleteAll
        @ltm.saveTasks
        write 'exit'
        exit
    end

    def inputHandler type, data
        case type
        when 'new'
            url = data['url']
            path = data['path']
            newTask path, url
        when 'start'
            id = data['id'].to_i
            startTask id
        when 'suspend'
            id = data['id'].to_i
            suspendTask id
        when 'delete'
            id = data['id'].to_i
            deleteTask id
        when 'connect'
            ipAddr = data['ipAddr']
            connect ipAddr
        when 'closeSupporter'
            closeSupporter
        when 'fetchInfo'
            sendInfo
        when 'exit'
            finalize
        end
    end
end
